lib/kino/subscription_manager.ex

```diff
@@ -0,0 +1,95 @@
+defmodule Kino.SubscriptionManager do
+  @moduledoc false
+
+  # The primary process responsible for subscribing to
+  # and broadcasting input/control events.
+
+  use GenServer
+
+  @name __MODULE__
+
+  @type state :: %{
+          subscribers_by_topic: %{(topic :: term()) => {teraget :: pid(), receive_as :: term()}}
+        }
+
+  def cross_node_name() do
+    {@name, node()}
+  end
+
+  @doc """
+  Starts the manager
+  """
+  def start_link(opts \\ []) do
+    GenServer.start_link(__MODULE__, opts, name: @name)
+  end
+
+  @doc """
+  Subscribes the given process to events under `topic`.
+
+  All events are sent as `{:event, receive_as, info}`,
+  where `receive_as` is the given term.
+  """
+  @spec subscribe(term(), pid(), term()) :: :ok
+  def subscribe(topic, pid, receive_as) do
+    GenServer.cast(@name, {:subscribe, topic, pid, receive_as})
+  end
+
+  @doc """
+  Unsubscribes the given process from events under `topic`.
+  """
+  @spec unsubscribe(term(), pid()) :: :ok
+  def unsubscribe(topic, pid) do
+    GenServer.cast(@name, {:unsubscribe, topic, pid})
+  end
+
+  @impl true
+  def init(_opts) do
+    {:ok, %{subscribers_by_topic: %{}}}
+  end
+
+  @impl true
+  def handle_cast({:subscribe, topic, pid, receive_as}, state) do
```

Comment:
Here `topic` is the input/control id that we subscribe to. We monitor the subscriber and remove it accordingly when it goes down, however we don't detect when the input/control (topic) no longer exists.

With processes it may not be _as_ critical, since usually we would expect the subscribing process to die on reevaluation, but once we start supporting functions we would keep those forever.

So it would be nice to detect when the inputs are removed and one solution I see is for the session to look at the outputs it removes and if there are any controls/inputs it sends a termination event to `destination`, just we do with regular events. WDYT?